zephyr/include/sof/lib/regions_mm.h

```diff
@@ -10,31 +10,122 @@
 
 #include <adsp_memory.h>
 #include <adsp_memory_regions.h>
+#include <ipc/topology.h>
+#include <rtos/alloc.h>
+#include <sof/list.h>
 #include <zephyr/drivers/mm/system_mm.h>
-#include <zephyr/sys/mem_blocks.h>
 #include <zephyr/init.h>
-#include <ipc/topology.h>
+#include <zephyr/sys/mem_blocks.h>
 
-/*
- * Struct containing information on virtual memory heap.
- * Information about allocated physical blocks is stored in
- * physical_blocks_allocators variable and uses zephyr memblocks api.
+/* Dependency on ipc/topology.h created due to memory capability definitions
+ * that are defined there
+ */
+
+/* API is NOT re-entry safe.
+ * This is due to our assumption that only management code will ever handle
+ * memory operations on heaps themselves.
+ */
```

Comment:
There are spinlocks in place in the memblocks api bitarrays.

Reply:
> There are spinlocks in place in the memblocks api bitarrays.

ok, so is this statement correct ? This code should not call itself (or will deadlock), but it should be safe to be called by any thread/core given use of splinlocks.

Reply:
No, the statement is this api is not designed to be used by multiple threads from the same core (ie freeing and simultaneously allocating on the same heap ) and the sharing of memory heap objects need to be controlled on the management code level (usage of shared heap).

Reply:
> No, the statement is this api is not designed to be used by multiple threads from the same core (ie freeing and simultaneously allocating on the same heap ) and the sharing of memory heap objects need to be controlled on the management code level (usage of shared heap).

Ok, this also implies it's not multicore capable ? This code needs to be thread/core safe.

Reply:
It is designed in a way that any multi thread operations should be controlled by the heap creator.

One of requirements was  that modules could create their own mini heaps as I understood. If the module is on the non zero core then it has to be able to do it. And it is responsibility of heap creator/client to handle access to it.

Secondly core id is used only for the "core heaps" creation. Those instances are designed not to be shared across cores and to be used as core specific memory. 

Only multi core capability/accessibility should be in case of "shared heap" and again the creator of the heap has to be responsible for its access.

@mmaka1 As far as I remember those were bullet points of our conversation or did i misunderstood the design ?

Reply:
> It is designed in a way that any multi thread operations should be controlled by the heap creator.

nak - each core API (and this is a core API) needs to be responsible for it's own thread/core safety and not rely on the layers above. Shared resource locking should be good enough in most cases here to fix this.

Reply:
Marking this as resolved since we agreed at the offline meeting that this API will only be used as managment code. Shared heap usage should be droped and shared region removed from Zephyr side.